### ISSUE TITLE
DefaultRegionChain with profile_name does not appear to check source_profile for a region

### ISSUE THREAD

**Issue by quodlibetor** [awslabs/aws-sdk-rust#443]
### What is the problem?

Using a credentials file that looks like:

```ini
[profile credentials]
aws_access_key_id = "..."
aws_secret_access_key = "..."
aws_session_token = "..."
region = us-east-1

[profile needs-source]
source_profile = credentials
role_arn = "..."

[profile needs-source-inline-region]
source_profile = credentials
role_arn = "..."
region = us-east-1
```

and a `main.rs` that looks like:

```rust
use std::env;

use aws_config::default_provider::credentials::DefaultCredentialsChain;
use aws_config::default_provider::region::DefaultRegionChain;

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>>{
    let profile = env::args().skip(1).next().expect("profile to use is required");

    let region = DefaultRegionChain::builder().profile_name(&profile).build().region().await;

    let creds = DefaultCredentialsChain::builder()
        .profile_name(&profile)
        .region(region.clone())
        .build()
        .await;

    let config = aws_config::from_env().credentials_provider(creds).region(region).load().await;

    let client = aws_sdk_sts::Client::new(&config);
    let _identity = client.get_caller_identity().send().await?;

    Ok(())
}
```

the following profile fails to load with a `NoRegion` error: `cargo run -- needs-source` but the other two do work.

Executing the aws cli with any of the profiles in that file succeeds in loading the profile.

This is a zip file that demonstrates all three cases: [repro.zip](https://github.com/awslabs/aws-sdk-rust/files/8027826/repro.zip)

### Version

```text
$ cargo tree | rg aws-
├── aws-config v0.3.0
│   ├── aws-http v0.3.0
│   │   ├── aws-smithy-http v0.33.1
│   │   │   ├── aws-smithy-types v0.33.1
│   │   ├── aws-smithy-types v0.33.1 (*)
│   │   ├── aws-types v0.3.0
│   │   │   ├── aws-smithy-async v0.33.1
│   │   │   ├── aws-smithy-types v0.33.1 (*)
│   ├── aws-sdk-sts v0.3.0
│   │   ├── aws-endpoint v0.3.0
│   │   │   ├── aws-smithy-http v0.33.1 (*)
│   │   │   ├── aws-types v0.3.0 (*)
│   │   ├── aws-http v0.3.0 (*)
│   │   ├── aws-sig-auth v0.3.0
│   │   │   ├── aws-sigv4 v0.3.0
│   │   │   │   ├── aws-smithy-http v0.33.1 (*)
│   │   │   ├── aws-smithy-http v0.33.1 (*)
│   │   │   ├── aws-types v0.3.0 (*)
│   │   ├── aws-smithy-async v0.33.1 (*)
│   │   ├── aws-smithy-client v0.33.1
│   │   │   ├── aws-smithy-async v0.33.1 (*)
│   │   │   ├── aws-smithy-http v0.33.1 (*)
│   │   │   ├── aws-smithy-http-tower v0.33.1
│   │   │   │   ├── aws-smithy-http v0.33.1 (*)
│   │   │   ├── aws-smithy-types v0.33.1 (*)
│   │   ├── aws-smithy-http v0.33.1 (*)
│   │   ├── aws-smithy-http-tower v0.33.1 (*)
│   │   ├── aws-smithy-query v0.33.1
│   │   │   ├── aws-smithy-types v0.33.1 (*)
│   │   ├── aws-smithy-types v0.33.1 (*)
│   │   ├── aws-smithy-xml v0.33.1
│   │   ├── aws-types v0.3.0 (*)
│   ├── aws-smithy-async v0.33.1 (*)
│   ├── aws-smithy-client v0.33.1 (*)
│   ├── aws-smithy-http v0.33.1 (*)
│   ├── aws-smithy-http-tower v0.33.1 (*)
│   ├── aws-smithy-json v0.33.1
│   │   └── aws-smithy-types v0.33.1 (*)
│   ├── aws-smithy-types v0.33.1 (*)
│   ├── aws-types v0.3.0 (*)
├── aws-sdk-sts v0.3.0 (*)
```

### Platform

macos and linux

### AWS Services

None.

**Comment by rcoh** [comment]
great catch! We'll get a fix out to match the behavior with the CLI.

**Comment by Velfi** [comment]
We merged a fix for this in [this PR](https://github.com/awslabs/smithy-rs/pull/1183). This fix will go out in the next release. If you like, you can target the `next` branch to try it out sooner:

```toml
[dependencies]
aws-config = { git = "https://github.com/awslabs/aws-sdk-rust", branch = "next" }
aws-sdk-sts = { git = "https://github.com/awslabs/aws-sdk-rust", branch = "next" }
```

**Comment by jdisanti** [comment]
The fix for this went out with `v0.7.0`.
